### PR TITLE
Support if statements in dataclass_transform class

### PR DESCRIPTION
**Pull request by KRunchPL**
Fixes: #14853

Adding support for `if` statements in the dataclass_transform class, so the attributes defined conditionally are treated as those that are directly in class body.

### CODE REVIEW

File: mypy/plugins/dataclasses.py [hunk 1]
```diff
@@ -380,6 +382,26 @@ def reset_init_only_vars(self, info: TypeInfo, attributes: list[DataclassAttribu
                             # recreate a symbol node for this attribute.
                             lvalue.node = None
 
+    @staticmethod
+    def _get_assignment_statements_from_if_statement(
+        stmt: IfStmt,
+    ) -> Generator[AssignmentStmt, None, None]:
```
Comment (sobolevn) [inline mypy/plugins/dataclasses.py:388]:
```suggestion
    ) -> Iterator[AssignmentStmt]:
```

File: mypy/plugins/dataclasses.py [hunk 2]
```diff
@@ -380,6 +382,26 @@ def reset_init_only_vars(self, info: TypeInfo, attributes: list[DataclassAttribu
                             # recreate a symbol node for this attribute.
                             lvalue.node = None
 
+    @staticmethod
```
Comment (sobolevn) [inline mypy/plugins/dataclasses.py:385]: Any reason to make this a `@staticmethod`? I think it only makes sense as an instance method. I would recommend to remove this decorator.

File: mypy/plugins/dataclasses.py [hunk 3]
```diff
@@ -380,6 +382,26 @@ def reset_init_only_vars(self, info: TypeInfo, attributes: list[DataclassAttribu
...
+    def _get_assignment_statements_from_if_statement(
+        stmt: IfStmt,
+    ) -> Generator[AssignmentStmt, None, None]:
+        for body in stmt.body:
+            if not body.is_unreachable:
+                yield from DataclassTransformer._get_assignment_statements_from_block(body)
+        if stmt.else_body is not None and not stmt.else_body.is_unreachable:
+            yield from DataclassTransformer._get_assignment_statements_from_block(stmt.else_body)
+
+    @staticmethod
```
Comment (sobolevn) [inline mypy/plugins/dataclasses.py:395]: And this.

File: test-data/unit/check-dataclass-transform.test [hunk 4]
```diff
@@ -451,3 +451,187 @@ Foo(1)  # E: Too many arguments for "Foo"
...
+class FunctionModel:
+    if TYPE_CHECKING:
+        present_1: int
+    else:
+        skipped_1: int
+    if True:
+        present_2: int
+    if not False:
+        present_3: int
+    if not TRUTH:
```
Comment (sobolevn) [inline test-data/unit/check-dataclass-transform.test:616]: Let's add at least a single test with some `def cond() -> bool: ...` function. 

```python
if cond():
   x: int
   y: int
   z1: int
else:
   x: str
   y: int
   z2: int
```

**Review by sobolevn — COMMENTED** [review]
Thanks! The usecase with `if TYPE_CHECKING` seems reasonable to me.
I personally use this from time to time.

File: mypy/plugins/dataclasses.py (hunk 1 above)
Comment (KRunchPL) [reply mypy/plugins/dataclasses.py]: Changed.

**Review by sobolevn — COMMENTED** [review]
One last thing from me: as far as I understand - this does also affect how `@dataclass`es are defined.

Can we please add a single test for it as well?

Thanks a lot for your work! 👍

File: mypy/plugins/dataclasses.py (hunk 3 above)
Comment (KRunchPL) [reply mypy/plugins/dataclasses.py]: Changed

**Comment by KRunchPL** [comment]
Added the test for pute dataclass.

File: test-data/unit/check-dataclass-transform.test (hunk 4 above)
Comment (sobolevn) [reply test-data/unit/check-dataclass-transform.test:616]: Yes, I think the test is correct 👍

**Review by sobolevn — APPROVED** [review]

**Comment by sobolevn** [comment]
Thank you!

**Pull request merged**